src/base/utils/tar.cpp

```diff
@@ -0,0 +1,298 @@
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "tar.h"
+
+#include <QByteArray>
+#include <QCoreApplication>
+#include <QHash>
+#include <QString>
+
+#include "base/logger.h"
+#include "base/utils/bytearray.h"
+#include "base/utils/fs.h"
+
+namespace
+{
+    const int TAR_HEADER_SIZE = 512;
+
+    const int TAR_FILENAME_OFFSET = 0;
+    const int TAR_FILESIZE_OFFSET = 124;
+    const int TAR_CHECKSUM_OFFSET = 148;
+    const int TAR_FILETYPE_OFFSET = 156;
+    const int TAR_FILENAME_PREFIX_OFFSET = 345;
+
+    const int TAR_FILENAME_SIZE = 100;
+    const int TAR_FILESIZE_SIZE = 12;
+    const int TAR_CHECKSUM_SIZE = 8;
+    const int TAR_FILENAME_PREFIX_SIZE = 155;
+
+    class TarEntry
+    {
+    public:
+        enum Types : unsigned char
+        {
+            REG_FILE_NULL = '\0',
+            REG_FILE_ASCII = '0',
+            // We don't care about the other file types
+
+            INVALID = std::numeric_limits<unsigned char>::max()
+        };
+
+        TarEntry() = default;
+
+        explicit TarEntry(const QByteArray &header)
+            : m_header(header)
+        {
+            if (m_header.size() != TAR_HEADER_SIZE)
+                return;
+
+            // Invalid header: The first header field represents the filename.
+            // If the 1st byte is null then we can't extract a name. So the whole header becomes invalid to us.
+            m_valid = !m_header.startsWith('\0');
+            verifyAndCheckNull();
+        }
+
+        QString name() const
+        {
+            if (!isValid())
+                return {};
+
+            const QByteArray filename = Utils::ByteArray::midView(m_header, TAR_FILENAME_OFFSET, TAR_FILENAME_SIZE);
+            const int filenameLen = strlen(filename);
+            if (filenameLen == 0)
+                return {};
+
+            const QByteArray filenamePrefix = Utils::ByteArray::midView(m_header, TAR_FILENAME_PREFIX_OFFSET, TAR_FILENAME_PREFIX_SIZE);
+            const int filenamePrefixLen = strlen(filenamePrefix);
+
+            const QString finalFilename = (filenamePrefixLen > 0) ?
+                                        QString::fromUtf8(filenamePrefix.constData(), filenamePrefixLen)
+                                              + '/' + QString::fromUtf8(filename.constData(), filenameLen)
+                                      : QString::fromUtf8(filename.constData(), filenameLen);
+
+            return Utils::Fs::toValidFileSystemName(finalFilename, true);
+        }
+
+        bool isNull() const { return m_null; }
+        bool isValid() const { return m_valid; }
+        bool isVerified() const { return m_verified; }
```

Comment:
Regarding `m_null`, `m_valid`, `m_verified`. Does these status intersect? I mean is it possible to have any two of them to be true at the same time?
If not it is better to tie them together into a separate type.

Reply:
Maybe I could merge `m_valid` and `m_verified`. Now `m_valid == true` only when the first byte isn't null. And `m_verified == true` when the checksum is correct. This allows to have more specific error output. But I could merge those 2 so `m_valid == true` only when 1st byte != null AND checksum is correct.

`m_null == true` when all bytes are null. It is used a special marker for EOF. I don't think I can combine it with the other 2 bools.

Reply:
Header can be either Invalid, Null or Regular so you just need enum with 3 values.